x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/ShapeDocValuesQuery.java

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.spatial.index.mapper;
+
+import org.apache.lucene.document.ShapeField;
+import org.apache.lucene.geo.Component2D;
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.index.DocValues;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.ConstantScoreScorer;
+import org.apache.lucene.search.ConstantScoreWeight;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.TwoPhaseIterator;
+import org.apache.lucene.search.Weight;
+import org.elasticsearch.xpack.spatial.index.fielddata.Component2DVisitor;
+import org.elasticsearch.xpack.spatial.index.fielddata.CoordinateEncoder;
+import org.elasticsearch.xpack.spatial.index.fielddata.GeometryDocValueReader;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/** Lucene geometry query for {@link BinaryShapeDocValuesField}. */
+abstract class ShapeDocValuesQuery<GEOMETRY> extends Query {
+
+    private final String field;
+    private final CoordinateEncoder encoder;
+    private final ShapeField.QueryRelation relation;
+    private final GEOMETRY[] geometries;
+
+    ShapeDocValuesQuery(String field, CoordinateEncoder encoder, ShapeField.QueryRelation relation, GEOMETRY[] geometries) {
+        if (field == null) {
+            throw new IllegalArgumentException("field must not be null");
+        }
+        this.field = field;
+        this.encoder = encoder;
+        this.relation = relation;
+        this.geometries = geometries;
+    }
+
+    protected abstract Component2D create(GEOMETRY geometry);
+
+    protected abstract Component2D create(GEOMETRY[] geometries);
```

Comment:
this is nice, do we need two methods? would not be enough with something like:
```
protected abstract Component2D create(GEOMETRY... geometry);
```

Reply:
That is what I did originally, but it turns out varargs does not play with generics, so in the generic class you have to use the two methods to get the same effect.

Reply:
I think it is because with type erasure we cannot tell the difference between Object[] and Object (since Object[] is also an Object).